### PR TITLE
chore: refactored the PaymentComponent, PresentableComponent, LoadingComponent, and created PaymentAwareComponent to follow the interface segregation rule.

### PR DESCRIPTION
**Pull request by mohammedDehairy**
chore: refactored the PaymentComponent, PresentableComponent, LoadingComponent, and created PaymentAwareComponent to follow the interface segregation rule.

### CODE REVIEW

**Review by descorp — APPROVED** [review]

File: AdyenDropIn/DropInComponent.swift [hunk 1]
```diff
@@ -14,7 +14,10 @@ import Adyen
 import UIKit
 
 /// A component that handles the entire flow of payment selection and payment details entry.
-public final class DropInComponent: NSObject, PresentableComponent {
+public final class DropInComponent: NSObject,
+    PresentableComponent,
+    PaymentAwareComponent,
+    LoadingComponent {
```
Comment (descorp) [inline AdyenDropIn/DropInComponent.swift:20]: We can probably remove this. Loading state of DropIn controlled by DropIn itself

File: Adyen/Components/Base/PaymentComponent.swift [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
+// Copyright (c) 2021 Adyen N.V.
 //
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
 import Foundation
 
 /// A component that handles the initial phase of getting payment details to initiate a payment.
-public protocol PaymentComponent: Component {
+public protocol PaymentComponent: PaymentAwareComponent {
```
Comment (descorp) [inline Adyen/Components/Base/PaymentComponent.swift:10]: Not sure if all `PaymentComponents` are `PaymentAwareComponent` 🤔

File: AdyenDropIn/Components/PaymentMethodListComponent.swift [hunk 3]
```diff
@@ -9,13 +9,48 @@ import Foundation
...
+        completion?()
+    }
+
+    /// :nodoc:
+    public func startLoading(for component: PaymentComponent) { /* Empty default implementation */ }
 }
 
 /// A component that presents a list of items for each payment method with a component.
-internal final class PaymentMethodListComponent: LoadingComponent, Localizable {
+internal final class PaymentMethodListComponent: LoadingComponent, PresentableComponent, Localizable {
```
Comment (descorp) [inline AdyenDropIn/Components/PaymentMethodListComponent.swift:53]: Isn't any `PresentableComponent` also a `Localizable` ?
Except for ApplePay maybe..

**Review by descorp — APPROVED** [review]

File: Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift [hunk 4]
```diff
@@ -9,7 +9,7 @@ import UIKit
 
 /// An abstract class that needs to be subclassed to abstract away any component
 /// whoes form consists of a combination of personal information pieces like first name, last name, phone, email, and billing address.
-open class AbstractPersonalInformationComponent: PaymentComponent, PresentableComponent, Localizable {
+open class AbstractPersonalInformationComponent: PaymentComponent, PresentableComponent, Localizable, LoadingComponent {
```
Comment (descorp) [inline Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift:12]: There is no `stopLoading` in this class

File: Adyen/Components/Base/LoadingComponent.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    /// Stops any processing animation that the view controller is running.
+    ///
+    /// - Parameters:
+    ///   - success: Boolean indicating the component should go to a success or failure state.
+    ///   - completion: Completion block to be called when animations are finished.
+    func stopLoading(withSuccess success: Bool, completion: (() -> Void)?)
+}
+
+extension LoadingComponent {
```
Comment (descorp) [inline Adyen/Components/Base/LoadingComponent.swift:22]: Do we really need a default implementation for this? It do nothing.
Better remove the extension and actually see where are using it :)

File: Demo/Common/IntegrationExample.swift [hunk 6]
```diff
@@ -86,13 +86,21 @@ internal final class IntegrationExample {
...
 
-        currentComponent?.stopLoading(withSuccess: false) { [weak self] in
+        stopLoadingIfNeeded(false) { [weak self] in
             self?.presenter?.dismiss { [weak self] in
                 if !isCancelled { self?.presentAlert(with: error) }
             }
         }
     }
 
+    private func stopLoadingIfNeeded(_ success: Bool, completion: (() -> Void)?) {
```
Comment (descorp) [inline Demo/Common/IntegrationExample.swift:96]: I have a feeling that `stopLoadingIfNeeded` is redundant: 
the `success: Bool` is never used, so it is basically a `Cancelable` with `didCancel(completion: (() -> Void)?)`

Comment (mohammedDehairy) [reply Demo/Common/IntegrationExample.swift:96]: `success: Bool` is used, not sure what you mean? not sure even what you mean by "redundant"? cancellable is when the user cancel the component, this is after the payment is concluded by success or failure.

File: Adyen/Components/Base/LoadingComponent.swift (hunk 5 above)
Comment (mohammedDehairy) [reply Adyen/Components/Base/LoadingComponent.swift]: removed the default implementation and made some more small refactoring of `LoadingComponent`.

File: Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift (hunk 4 above)
Comment (mohammedDehairy) [reply Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift]: its in an extension, moved the protocol conformance to the extension. 👍

File: Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponentExtensions.swift [hunk 7]
```diff
@@ -18,7 +18,7 @@ extension AbstractPersonalInformationComponent: FormViewControllerDelegate {
     }
 }
 
-extension AbstractPersonalInformationComponent {
+extension AbstractPersonalInformationComponent: LoadingComponent {
```
Comment (descorp) [inline Adyen/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponentExtensions.swift:21]: 👌

**Review by descorp — APPROVED** [review]

**Pull request merged**